core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala

```diff
@@ -220,7 +220,8 @@ private[spark] class MemoryStore(
     }
 
     // Unroll this block safely, checking whether we have exceeded our threshold periodically
-    while (values.hasNext && keepUnrolling) {
+    // and if no thread interrupts have been received.
+    while (values.hasNext && keepUnrolling && !Thread.currentThread().isInterrupted) {
```

Comment:
I am slightly wary of calling `isInterrupted` for every iteration.
Given we already have the `if  (elementsUnrolled ...` condition below, do we want to move it to that ?

IIRC `isInterrupted` is a couple of orders more expensive than a boolean check.

Something like:
```
var interrupted = false
while (values.hasNext && keepUnrolling && !interrupted) {
  valueHolder ...
  if (... ){
    interrupted = Thread.currentThread().isInterrupted
   ...
}
```

Reply:
@mridulm - if we put this check inside the `if` condition below, it seems we will only set the `interrupted` flag the next time the memory check period expires ? Seems like the default value for the check period is set to 16. So, if the interrupt is received after the first element is unrolled, we will wait for 15 more elements and also do extra work in the interim, that we probably have to dispose later anyway ? So might be better to check within the `while` loop and exit earlier as we are doing in the PR ?

```
      valuesHolder.storeValue(values.next())
      if (elementsUnrolled % memoryCheckPeriod == 0) {
        val currentSize = valuesHolder.estimatedSize()
        // If our vector's size has exceeded the threshold, request more memory
```

Reply:
> IIRC isInterrupted is a couple of orders more expensive than a boolean check.

In older JDKs it is implemented via an intrinsic and in newer ones (14+) it's just a volatile boolean check (see https://bugs.openjdk.org/browse/JDK-8229516), so I don't think it's too expensive. There's an old StackOverflow answer at https://stackoverflow.com/a/5158441/590203 offering a plausible explanation for why the intrinsic should be cheap.

On JDK 11 on my laptop, I ran a toy JMH benchmark and measured < 1ns per call. 

I think these costs are very small in comparison to the per-element serialization costs.

Given this, I'm not too worried about performance regressions due to this change and think it's probably okay to check on every element rather than every 16th.

Reply:
@anishshri-db Thread interruption is a rare occurrence when compared to number of `putIterator` calls - so we should try to minimize the impact for the common case, as long as it is a reasonable additional cost when interruption does occur (having said that, please see more below).

@JoshRosen It is way faster than I expected it to be [1] - while it is still slower than a local variable [2], the cost is really low when compared to the cost of other operations in the loop ... this should be just noise.

Given this, while in general I prefer to minimize unnecessary cost, I am fine with the change.


[1] Thanks for your comments, went digging into this - was fun !
I knew it was intrinsic, but the uncontended reads are much faster than I had expected (I seem to be misremembering some stats) - and thanks for jdk14 ref, was not aware of that change.

[2] 
On my linux desktop: 2.175 +- 0.001 ns/op versus 2.424 +- 0.001 ns/op